File: ReactAndroid/src/main/java/com/facebook/systrace/Systrace.java
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.systrace;

import android.os.Trace;

/**
 * Systrace stub that mostly does nothing but delegates to Trace for beginning/ending sections. The
 * internal version of this file has not been opensourced yet.
 */
public class Systrace {

  public static final long TRACE_TAG_REACT_JAVA_BRIDGE = 0L;
  public static final long TRACE_TAG_REACT_APPS = 0L;
  public static final long TRACE_TAG_REACT_FRESCO = 0L;
  public static final long TRACE_TAG_REACT_VIEW = 0L;
  public static final long TRACE_TAG_REACT_JS_VM_CALLS = 0L;

  public enum EventScope {
    THREAD('t'),
    PROCESS('p'),
    GLOBAL('g');

    private final char mCode;

    private EventScope(char code) {
      mCode = code;
    }

    public char getCode() {
      return mCode;
    }
  }

  public static void registerListener(TraceListener listener) {}

  public static void unregisterListener(TraceListener listener) {}

  public static boolean isTracing(long tag) {
    return false;
  }

  public static void traceInstant(long tag, final String title, EventScope scope) {}

  public static void beginSection(long tag, final String sectionName) {
    Trace.beginSection(sectionName);
  }

  public static void endSection(long tag) {
    Trace.endSection();
  }

  public static void beginAsyncSection(long tag, final String sectionName, final int cookie) {}

  public static void beginAsyncSection(
      long tag, final String sectionName, final int cookie, final long startNanos) {}

  public static void endAsyncSection(long tag, final String sectionName, final int cookie) {}

  public static void endAsyncSection(
      long tag, final String sectionName, final int cookie, final long endNanos) {}

  public static void traceCounter(long tag, final String counterName, final int counterValue) {}

  public static void startAsyncFlow(long tag, final String sectionName, final int cookie) {}

  public static void stepAsyncFlow(long tag, final String sectionName, final int cookie) {}

  public static void endAsyncFlow(long tag, final String sectionName, final int cookie) {}
}

File: packages/react-native/ReactAndroid/src/main/java/com/facebook/systrace/Systrace.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.systrace

import androidx.tracing.Trace
import java.lang.Runnable
import kotlin.text.StringBuilder

/**
 * Systrace stub that mostly does nothing but delegates to Trace for beginning/ending sections. The
 * internal version of this file has not been opensourced yet.
 */
@Suppress("UNUSED_PARAMETER")
internal object Systrace {

  const val TRACE_TAG_REACT: Long = 0L

  @JvmStatic fun registerListener(listener: TraceListener?): Unit = Unit

  @JvmStatic fun unregisterListener(listener: TraceListener?): Unit = Unit

  @JvmStatic fun isTracing(tag: Long): Boolean = false

  @JvmStatic fun traceInstant(tag: Long, title: String?, scope: EventScope?): Unit = Unit

  @JvmStatic
  fun traceSection(tag: Long, sectionName: String, block: Runnable) {
    beginSection(tag, sectionName)
    try {
      block.run()
    } finally {
      endSection(tag)
    }
  }

  @JvmStatic
  fun beginSection(tag: Long, sectionName: String) {
    Trace.beginSection(sectionName)
  }

  @JvmStatic
  fun beginSection(tag: Long, sectionName: String, args: Array<String>, argsLength: Int) {
    Trace.beginSection(sectionName + "|" + convertArgsToText(args, argsLength))
  }

  private fun convertArgsToText(args: Array<String>, argsLength: Int): String {
    val argsText: StringBuilder = StringBuilder()
    var ii = 1
    while (ii < argsLength) {
      val key = args[ii - 1]
      val value = args[ii]
      argsText.append(key)
      argsText.append('=')
      argsText.append(value)
      if (ii < argsLength - 1) {
        argsText.append(';')
      }
      ii += 2
    }
    return argsText.toString()
  }

  @JvmStatic
  fun endSection(tag: Long) {
    Trace.endSection()
  }

  @JvmStatic
  fun beginAsyncSection(tag: Long, sectionName: String, cookie: Int) {
    Trace.beginAsyncSection(sectionName, cookie)
  }

  @JvmStatic
  fun beginAsyncSection(tag: Long, sectionName: String, cookie: Int, startNanos: Long) {
    beginAsyncSection(tag, sectionName, cookie)
  }

  @JvmStatic
  fun endAsyncSection(tag: Long, sectionName: String, cookie: Int) {
    Trace.endAsyncSection(sectionName, cookie)
  }

  @JvmStatic
  fun endAsyncSection(tag: Long, sectionName: String, cookie: Int, endNanos: Long) {
    endAsyncSection(tag, sectionName, cookie)
  }

  @JvmStatic
  fun traceCounter(tag: Long, counterName: String, counterValue: Int) {
    Trace.setCounter(counterName, counterValue)
  }

  @JvmStatic
  fun startAsyncFlow(tag: Long, sectionName: String, cookie: Int) {
    beginAsyncSection(tag, sectionName, cookie)
  }

  @JvmStatic fun stepAsyncFlow(tag: Long, sectionName: String, cookie: Int): Unit = Unit

  @JvmStatic
  fun endAsyncFlow(tag: Long, sectionName: String, cookie: Int) {
    endAsyncSection(tag, sectionName, cookie)
  }

  enum class EventScope(val code: Char) {
    THREAD('t'),
    PROCESS('p'),
    GLOBAL('g'),
  }
}
